File: 1091_Shortest Path in Binary Matrix/ShortestPathinBinaryMatrix.cs
public class Solution {
    public int ShortestPathBinaryMatrix(int[][] grid) {
        var m = grid.Length;
        if(m == 0) return -1;
        var n = grid[0].Length;
        if(grid[0][0] != 0) return -1;
        if(grid[m-1][n-1] != 0) return -1;
        if(m == 1 && n == 1) return 1;
        
         var ans = 0;
        var queue = new Queue<int[]>();
        queue.Enqueue(new int[]{0,0});
        grid[0][0] = 1;
        
        var dirs_x = new int[]{-1,-1, 0, 1,1,1,0,-1};
        var dirs_y = new int[]{ 0,-1,-1,-1,0,1,1, 1};
        
        while(queue.Count() > 0)
        {
            ans++;
            var c = queue.Count();
            while(c-- > 0)
            {
                var pos = queue.Dequeue();
                for(int i=0;i<8;i++)
                {
                    var dx = pos[0] + dirs_x[i];
                    var dy = pos[1] + dirs_y[i];
                    
                    if(dx < 0 || dy < 0 || dx == m || dy == n || grid[dx][dy] != 0) continue;
                    if(dx == m-1 && dy == n-1) return ans+1;
                    grid[dx][dy] = 1;
                    queue.Enqueue(new int[]{dx, dy});
                }
            }
        }
        
        return -1;
    }
}

File: 1091_Shortest Path in Binary Matrix/ShortestPathinBinaryMatrix.java
class Solution {
    public int shortestPathBinaryMatrix(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] dirs = new int[8][];
        dirs[0] = new int[] {-1, 1};
        dirs[1] = new int[] {-1, 0};
        dirs[2] = new int[] {-1,-1};
        dirs[3] = new int[] {0,  1};
        dirs[4] = new int[] {0, -1};
        dirs[5] = new int[] {1, -1};
        dirs[6] = new int[] {1,  0};
        dirs[7] = new int[] {1,  1};
        
        if(grid[0][0] == 1 || grid[m-1][n-1] == 1) return -1;
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0,0});
        grid[0][0] = 1;
        
        int ans = 0;
        while(queue.size() > 0) {
            int size = queue.size();
            ans++;
            while(size-- > 0) {
                int[] pos = queue.poll();
                int x = pos[0], y = pos[1];    
                if(x+1 == m && y+1 == n) return ans;
                
                for(int i=0;i< dirs.length;i++) {
                    int dx = dirs[i][0] + x;
                    int dy = dirs[i][1] + y;
                    if(dx < 0 || dy < 0 || dx == m || dy == n || grid[dx][dy] != 0) continue;
                    queue.offer(new int[]{dx, dy});
                    grid[dx][dy] = 1;
                }
            }
        }
        
        return -1;
    }
}